File: pkg-py/tests/test_prompt.py
"""Prompt rendering, including the cases shared with the R package."""

import re

import jinja2
import pandas as pd
import pytest

from commons import data_source
from commons._catalog._core import Manifest, Relation
from commons._data_dictionary import DataDictionary
from commons._data_source import DataSource, TableId
from commons._definitions._registry import build_registry
from commons._prompt import (
    EXECUTION_TOOL,
    check_instructions,
    citation_trust_exception,
    is_claude_5_model,
    prompt_date,
    read_instructions,
    read_prompt,
    render_system_prompt,
    system_prompt_data,
    system_prompt_template,
    tool_availability,
)

from ._shared import SHARED_DIR, load_shared_fixture

PROMPT_DATA = load_shared_fixture("prompt-data")


def _own_execution_tool(tool: str) -> str:
    """Resolve the fixtures' `$execution_tool` placeholder to this package's."""
    return EXECUTION_TOOL if tool == "$execution_tool" else tool

_PAD = re.compile(r"^pad:(\d+)$")


def _expand_pads(spec):
    """Grow the `pad:<n>` labels the fixture uses to exceed the index cap."""
    if isinstance(spec, dict):
        return {key: _expand_pads(value) for key, value in spec.items()}
    if isinstance(spec, list):
        return [_expand_pads(item) for item in spec]
    if isinstance(spec, str):
        padded = _PAD.match(spec)
        return "x" * int(padded.group(1)) if padded else spec
    return spec


def _manifest(catalog) -> Manifest:
    labels = [
        f"{catalog['label_prefix']}{index}"
        for index in range(1, catalog["objects"] + 1)
    ]
    relations = {label: Relation(id=TableId(table=label)) for label in labels}
    manifest = Manifest.build(relations, namespace_selected=True)
    assert manifest.searchable == catalog["searchable"]
    return manifest


def _hydrate(spec) -> DataSource:
    dictionary = spec.get("dictionary")
    source = data_source(
        dictionary=None
        if dictionary is None
        else DataDictionary.model_validate(_expand_pads(dictionary)),
        **{name: pd.DataFrame({"n": [1]}) for name in spec["tables"]},
    )
    if spec.get("catalog") is not None:
        source.manifest = _manifest(spec["catalog"])
    return source


def test_shared_render_cases():
    fixture = load_shared_fixture("prompt-render")
    cases = fixture["render"]["cases"]
    assert cases

    template = system_prompt_template()
    for case in cases:
        expected = (
            (SHARED_DIR / case["expected"]).read_text(encoding="utf-8").rstrip("\n")
        )
        assert render_system_prompt(template, case["data"]) == expected, case["name"]


def test_shared_tool_data_cases():
    cases = load_shared_fixture("prompt-render")["tool_data"]["cases"]
    assert cases

    for case in cases:
        tools = [_own_execution_tool(tool) for tool in case["tools"]]
        data: dict[str, object] = dict(tool_availability(tools))
        data["citation_trust_exception"] = citation_trust_exception(tools)
        got = {key: data[key] for key in case["expected"]}
        assert got == case["expected"], case["name"]


def test_shared_template_render_cases():
    cases = load_shared_fixture("prompt-render")["template_render"]["cases"]
    assert cases

    for case in cases:
        got = render_system_prompt(case["template"], case["data"])
        assert got == case["expected"], case["name"]


def test_shared_unsupported_templates():
    cases = load_shared_fixture("prompt-render")["unsupported"]["cases"]
    assert cases

    for case in cases:
        with pytest.raises(ValueError, match="Unsupported prompt template syntax"):
            render_system_prompt(case["template"], case["data"])


def test_shared_malformed_templates():
    cases = load_shared_fixture("prompt-render")["malformed"]["cases"]
    assert cases

    for case in cases:
        with pytest.raises((ValueError, TypeError, jinja2.TemplateSyntaxError)):
            render_system_prompt(case["template"], case["data"])


def test_shared_rejected_data():
    cases = load_shared_fixture("prompt-render")["rejected_data"]["cases"]
    assert cases

    for case in cases:
        with pytest.raises((TypeError, ValueError)):
            render_system_prompt(case["template"], case["data"])


def test_conditional_sections():
    template = (
        "<!-- source-only note -->\n"
        "{% if enabled %}\n"
        "{% if nested %}\nEnabled\nNested\n{% else %}\nEnabled\nNot nested\n{% endif %}\n"
        "{% else %}\nDisabled\n{% endif %}"
    )

    assert render_system_prompt(template, {"enabled": True, "nested": False}) == (
        "Enabled\nNot nested"
    )
    assert (
        render_system_prompt(template, {"enabled": False, "nested": True}) == "Disabled"
    )
    assert (
        render_system_prompt("{% if not on %}\nOff\n{% endif %}", {"on": False})
        == "Off"
    )


def test_values_are_not_rendered_again():
    template = (
        "Tables:\n{{ tables }}\nUse `{{ definition_token }}`. "
        "Write a literal {% raw %}`{{name}}`{% endraw %} token."
    )
    data = {
        "tables": "- sales\\daily\n- orders {{raw}}",
        "definition_token": "{{name}}",
    }

    assert render_system_prompt(template, data) == (
        "Tables:\n- sales\\daily\n- orders {{raw}}\nUse `{{name}}`. "
        "Write a literal `{{name}}` token."
    )


def test_template_data_is_validated():
    with pytest.raises(ValueError, match="has no value"):
        render_system_prompt("{% if unknown %}\nx\n{% endif %}", {})
    with pytest.raises(TypeError, match="must be True or False"):
        render_system_prompt("{% if tables %}\nx\n{% endif %}", {"tables": "a"})
    with pytest.raises(TypeError, match="must be a string"):
        render_system_prompt("{{ tables }}", {"tables": ["a", "b"]})


def test_claude_5_model_ids_are_recognized_across_providers():
    assert is_claude_5_model("claude-sonnet-5")
    assert is_claude_5_model("anthropic/claude-opus-5")
    assert is_claude_5_model("us.anthropic.claude-fable-5")
    assert is_claude_5_model("databricks-claude-sonnet-5")
    assert not is_claude_5_model("claude-sonnet-4-5")
    assert not is_claude_5_model("gpt-5.4")
    assert not is_claude_5_model(None)


def test_missing_instruction_paths_are_recognized(tmp_path):
    with pytest.raises(FileNotFoundError):
        check_instructions("missing-instructions.Rmd")
    with pytest.raises(FileNotFoundError):
        check_instructions("missing-instructions.template")
    with pytest.raises(FileNotFoundError):
        check_instructions("missing-dir/instructions")
    check_instructions("Be concise.")
    check_instructions("Line one.\nLine two.")
    check_instructions(None)

    path = tmp_path / "instructions.md"
    path.write_text("Prefer weekly grain.\n", encoding="utf-8")
    assert read_instructions(str(path)) == "Prefer weekly grain."
    assert read_instructions("Be concise.") == "Be concise."


def test_citation_request_text_names_the_dialect():
    reminder = read_prompt("citation-request.md")

    assert "<commons-citation>" in reminder
    assert "\n" not in reminder


@pytest.mark.parametrize("case", PROMPT_DATA["cases"], ids=lambda case: case["name"])
def test_shared_prompt_data_cases(case):
    sources = {spec["name"]: _hydrate(spec) for spec in case["sources"]}
    data = system_prompt_data(
        sources,
        build_registry(sources),
        instructions=case.get("instructions"),
        tools=[_own_execution_tool(tool) for tool in case.get("tools", [])],
        model=case.get("model"),
    )

    assert list(data) == PROMPT_DATA["fields"]
    assert (
        data["execution_tool"]
        == PROMPT_DATA["substitutions"]["python"]["execution_tool"]
    )
    assert {key: data[key] for key in case["expect"]} == case["expect"]


def test_shared_prompt_data_cases_are_not_empty():
    assert PROMPT_DATA["cases"]


def test_prompt_data_renders_the_packaged_template():
    source = data_source(orders=pd.DataFrame({"n": [1]}))
    sources = {"sales_db": source}
    data = system_prompt_data(sources, build_registry(sources), tools=["run_sql"])

    prompt = render_system_prompt(system_prompt_template(), data)

    assert data["date"] == prompt_date()
    assert "{{" not in prompt.replace("`{{name}}`", "")


def test_prompt_data_reads_tool_names_from_any_iterable():
    source = data_source(orders=pd.DataFrame({"n": [1]}))
    sources = {"sales_db": source}
    # Both the trust exception and the per-tool flags read `tools`, so a
    # one-shot iterator has to survive being read twice.
    data = system_prompt_data(
        sources, build_registry(sources), tools=iter(["call_metrics"])
    )

    assert data["has_call_metrics"]
    assert data["citation_trust_exception"].endswith("`call_metrics`")

File: pkg-r/tests/testthat/test-prompt.R
test_that("prompt templates select conditional sections", {
  template <- paste(
    "<!-- source-only note -->",
    "{% if enabled %}",
    "{% if nested %}",
    "Enabled",
    "Nested",
    "{% else %}",
    "Enabled",
    "Not nested",
    "{% endif %}",
    "{% else %}",
    "Disabled",
    "{% endif %}",
    sep = "\n"
  )

  expect_equal(
    render_system_prompt(template, list(enabled = TRUE, nested = FALSE)),
    "Enabled\nNot nested"
  )
  expect_equal(
    render_system_prompt(template, list(enabled = FALSE, nested = TRUE)),
    "Disabled"
  )
  expect_equal(
    render_system_prompt(
      "{% if not enabled %}\nOff\n{% endif %}",
      list(enabled = FALSE)
    ),
    "Off"
  )
})

test_that("prompt templates interpolate runtime data without recursion", {
  template <- paste(
    "Tables:\n{{ tables }}\nUse `{{ definition_token }}`.",
    "Write a literal {% raw %}`{{name}}`{% endraw %} token."
  )
  data <- list(tables = "- sales\\daily\n- orders {{raw}}")
  data$definition_token <- "{{name}}"

  expect_equal(
    render_system_prompt(template, data),
    paste(
      "Tables:\n- sales\\daily\n- orders {{raw}}\nUse `{{name}}`.",
      "Write a literal `{{name}}` token."
    )
  )
})

test_that("prompt templates validate their structure and values", {
  expect_error(
    render_system_prompt("{% if unknown %}\nx\n{% endif %}", list()),
    "which has no value"
  )
  expect_error(
    render_system_prompt("{% if yes %}\nx\n", list(yes = TRUE)),
    "unclosed"
  )
  expect_error(
    render_system_prompt("{% raw %}\nx\n", list()),
    "unclosed"
  )
  expect_error(
    render_system_prompt("{% endif %}", list()),
    "outside an"
  )
  expect_error(
    render_system_prompt("{% for x in y %}\n{% endfor %}", list()),
    "Unsupported prompt template syntax"
  )
  expect_error(
    render_system_prompt("{% if tables %}\nx\n{% endif %}", list(tables = "a")),
    "must be TRUE or FALSE"
  )
  expect_error(
    render_system_prompt("{{ tables }}", list(tables = c("a", "b"))),
    "must be a single string"
  )
})

test_that("both packages render the shared template cases the same way", {
  cases <- shared_fixture("prompt-render")$template_render$cases
  expect_gt(length(cases), 0)

  for (case in cases) {
    expect_identical(
      render_system_prompt(case$template, case$data),
      case$expected,
      info = case$name
    )
  }
})

test_that("both packages reject the same unsupported templates", {
  cases <- shared_fixture("prompt-render")$unsupported$cases
  expect_gt(length(cases), 0)

  for (case in cases) {
    expect_error(
      render_system_prompt(case$template, case$data),
      "Unsupported prompt template syntax",
      info = case$name
    )
  }
})

test_that("both packages reject the same malformed templates", {
  cases <- shared_fixture("prompt-render")$malformed$cases
  expect_gt(length(cases), 0)

  for (case in cases) {
    expect_error(
      render_system_prompt(case$template, case$data),
      regexp = NULL,
      info = case$name
    )
  }
})

test_that("both packages reject the same template data", {
  cases <- shared_fixture("prompt-render")$rejected_data$cases
  expect_gt(length(cases), 0)

  for (case in cases) {
    expect_error(
      render_system_prompt(case$template, case$data),
      regexp = NULL,
      info = case$name
    )
  }
})

test_that("both packages render the shared prompt cases the same way", {
  fixture <- shared_fixture("prompt-render")
  cases <- fixture$render$cases
  expect_gt(length(cases), 0)

  template <- read_system_prompt()
  for (case in cases) {
    expected <- paste(
      readLines(
        test_path("fixtures", "shared", case$expected),
        warn = FALSE,
        encoding = "UTF-8"
      ),
      collapse = "\n"
    )
    expect_equal(
      render_system_prompt(template, case$data),
      expected,
      info = case$name
    )
  }
})

test_that("tool availability and the trust exception follow the shared cases", {
  cases <- shared_fixture("prompt-render")$tool_data$cases
  expect_gt(length(cases), 0)

  for (case in cases) {
    tools <- as.character(unlist(case$tools))
    data <- c(
      tool_availability(tools),
      list(citation_trust_exception = citation_trust_exception(tools))
    )
    expect_equal(data[names(case$expected)], case$expected, info = case$name)
  }
})

test_that("missing instruction paths are recognized", {
  expect_error(
    check_instructions("missing-instructions.Rmd"),
    "does not exist"
  )
  expect_error(
    check_instructions("missing-instructions.template"),
    "does not exist"
  )
  expect_error(check_instructions("missing-dir/instructions"), "does not exist")
  expect_no_error(check_instructions("Be concise."))
  expect_no_error(check_instructions("Line one.\nLine two."))
  expect_no_error(check_instructions(NULL))
})

test_that("the packaged prompt leaves no template markup", {
  template <- read_system_prompt()
  prompt <- test_agent()$get_system_prompt()

  expect_match(template, "{{name}}", fixed = TRUE)
  expect_no_match(prompt, "<!--", fixed = TRUE)
  expect_no_match(prompt, "{{ date }}", fixed = TRUE)
  expect_no_match(prompt, "# Governed definitions", fixed = TRUE)
})

# The fixture writes a label meant to overflow the definition index as
# `pad:<n>`, so the file carries the length rather than the characters.
prompt_data_expand_pads <- function(spec) {
  if (is.list(spec)) {
    return(lapply(spec, prompt_data_expand_pads))
  }
  if (is.character(spec) && length(spec) == 1L && grepl("^pad:[0-9]+$", spec)) {
    return(strrep("x", as.integer(sub("^pad:", "", spec))))
  }
  spec
}

prompt_data_manifest <- function(catalog) {
  labels <- paste0(catalog$label_prefix, seq_len(catalog$objects))
  relations <- rep_len(list(list(kind = "table")), length(labels))
  names(relations) <- labels
  manifest <- new_catalog_manifest(relations, namespace_selected = TRUE)
  expect_identical(manifest$searchable, catalog$searchable)
  manifest
}

prompt_data_source <- function(spec) {
  tables <- as.character(unlist(spec$tables))
  frames <- rep_len(list(data.frame(n = 1)), length(tables))
  names(frames) <- tables
  dictionary <- if (is.null(spec$dictionary)) {
    NULL
  } else {
    new_data_dictionary(prompt_data_expand_pads(spec$dictionary))
  }
  source <- suppressMessages(
    rlang::exec(data_source, !!!frames, dictionary = dictionary)
  )
  if (!is.null(spec$catalog)) {
    # Private state is an environment, so the manifest a warehouse listing
    # would have left behind can be attached to a frame source in place.
    state <- data_source_state(source)
    state$manifest <- prompt_data_manifest(spec$catalog)
  }
  source
}

test_that("both packages derive the same prompt data from the same sources", {
  fixture <- shared_fixture("prompt-data")
  cases <- fixture$cases
  expect_gt(length(cases), 0)
  fields <- as.character(unlist(fixture$fields))

  for (case in cases) {
    sources <- lapply(case$sources, prompt_data_source)
    names(sources) <- vapply(case$sources, `[[`, character(1), "name")

    data <- system_prompt_data(
      sources,
      definitions_registry(sources),
      instructions = case$instructions,
      tools = as.character(unlist(case$tools)),
      model = case$model
    )

    expect_named(data, fields, info = case$name)
    expect_equal(data[names(case$expect)], case$expect, info = case$name)
  }
})

test_that("prompt data renders the packaged template", {
  sources <- list(sales_db = test_source())
  data <- system_prompt_data(
    sources,
    definitions_registry(sources),
    tools = "run_sql"
  )
  prompt <- render_system_prompt(read_system_prompt(), data)

  expect_identical(data$date, as.character(Sys.Date()))
  expect_no_match(
    gsub("`{{name}}`", "", prompt, fixed = TRUE),
    "{{",
    fixed = TRUE
  )
})

test_that("Claude 5 model IDs are recognized across providers", {
  expect_true(is_claude_5_model("claude-sonnet-5"))
  expect_true(is_claude_5_model("anthropic/claude-opus-5"))
  expect_true(is_claude_5_model("us.anthropic.claude-fable-5"))
  expect_true(is_claude_5_model("databricks-claude-sonnet-5"))
  expect_false(is_claude_5_model("claude-sonnet-4-5"))
  expect_false(is_claude_5_model("gpt-5.4"))
  expect_false(is_claude_5_model(NULL))
})

test_that("instructions are not interpreted as prompt template expressions", {
  instructions <- "Use `{tables}` exactly as written."
  prompt <- test_agent(instructions = instructions)$get_system_prompt()

  expect_true(endsWith(prompt, instructions))
})
